sokoban.hs

```diff
@@ -102,27 +132,98 @@ moveToGoal b (pr,pc) (r,c)
 
+-- given board, player coords, box coords, and box destination coords (being unoccupied space),
+-- returns next board state
+pushToUnoccupied :: Board -> Coordinates -> Coordinates -> Coordinates -> Board
+pushToUnoccupied b (pr,pc) (br,bc) (r,c)
+  | player == '@' = (writeToCell (writeToCell (writeToCell b (pr,pc) ' ') (br,bc) '@') (r,c) '$')
+  | player == '+' = (writeToCell (writeToCell (writeToCell b (pr,pc) '.') (br,bc) '@') (r,c) '$')
+  where player = getCharacter b (pr,pc)
+
+-- given board, player coords, box coords, and box destination coords (being a goal),
+-- returns next board state
+pushToGoal :: Board -> Coordinates -> Coordinates -> Coordinates -> Board
+pushToGoal b (pr,pc) (br,bc) (r,c)
+  | player == '@' = (writeToCell (writeToCell (writeToCell b (pr,pc) ' ') (br,bc) '@') (r,c) '*')
+  | player == '+' = (writeToCell (writeToCell (writeToCell b (pr,pc) '.') (br,bc) '@') (r,c) '*')
+  where player = getCharacter b (pr,pc)
+
+-- given board, player coords, box-on-goal coords, and box destination coords (being unoccupied space),
+-- returns next board state
+pushGoalToUnoccupied :: Board -> Coordinates -> Coordinates -> Coordinates -> Board
+pushGoalToUnoccupied b (pr,pc) (br,bc) (r,c)
+  | player == '@' = (writeToCell (writeToCell (writeToCell b (pr,pc) ' ') (br,bc) '+') (r,c) '$')
+  | player == '+' = (writeToCell (writeToCell (writeToCell b (pr,pc) '.') (br,bc) '+') (r,c) '$')
+  where player = getCharacter b (pr,pc)
+
+-- given board, player coords, box-on-goal coords, and box destination coords (being a goal),
+-- returns next board state
+pushGoalToGoal :: Board -> Coordinates -> Coordinates -> Coordinates -> Board
+pushGoalToGoal b (pr,pc) (br,bc) (r,c)
+  | player == '@' = (writeToCell (writeToCell (writeToCell b (pr,pc) ' ') (br,bc) '+') (r,c) '*')
+  | player == '+' = (writeToCell (writeToCell (writeToCell b (pr,pc) '.') (br,bc) '+') (r,c) '*')
+  where player = getCharacter b (pr,pc)
+
+-- given board,
+-- returns True if given state is a winning state, false otherwise
+-- (no '.' on board and player is not on a goal ('+'))
+isGameWon :: State -> Bool
+isGameWon (State board (pr,pc)) = (getCharacter board (pr,pc) /= '+') && not (foldr (\r b -> b || (elem '.' r)) False board)
+
 {-- Game Functions --}
 
+-- main sokoban function
+play :: Result -> IO()
+play (ContinueGame s) = 
+  do
+    printBoard (board s)
+    putStr "Input your next move (W,A,S,D):"
+    move <- getLine
+    case move of
+      [] -> play (ContinueGame s) -- continue if no input
+      _ -> play (getNextBoard (Action (toUpper (move!!0))) s)
+
+play (WonGame s) =
+  do
+    printBoard (board s)
+    putStrLn "Congratulations! You won the game!"
+
 -- takes user input and current state of board
--- returns the next state of board
+-- returns the next state of board as a result (game won or continue game)
 -- should be used as a helper in the main game function
-getNextBoard :: Action -> State -> State
+getNextBoard :: Game
```

Comment:
i was thinking that we would use this function to just get the next state, and then do the win condition checking (to create ContinueGame / EndGame, but this is fine too

Reply:
good work

Reply:
yeah I did that because not all cases can result in winning so there's no need to check every time
but actually it's just when a box is pushed from non-goal ($, not *) onto a goal so I'm checking more than I need to (which I'll fix)